File: host/zbsniff/proto.py
"""Wire protocol (v1) — Python mirror of firmware/src/core/codec.c.

Source of truth: /protocol/framing.md and /protocol/framing.json. This module
implements the same CRC, framing, and message layouts so the host and firmware
interoperate. Kept dependency-free for easy unit testing.
"""

from __future__ import annotations

import json
import struct
from dataclasses import dataclass
from enum import IntEnum

MAGIC0 = 0x5A
MAGIC1 = 0xBE
PROTO_VER = 1

MAX_PSDU = 127
MAX_PAYLOAD = 2048

CHANNEL_MIN = 11
CHANNEL_MAX = 26


class Msg(IntEnum):
    CAPTURED_FRAME = 0x01
    ED_RESULT = 0x02
    STATUS = 0x03
    LOG = 0x04
    ACK = 0x05
    INCIDENT = 0x06
    CMD_SET_CHANNEL = 0x81
    CMD_SET_MODE = 0x82
    CMD_START = 0x83
    CMD_STOP = 0x84
    CMD_SET_HOP = 0x85
    CMD_ED_SCAN = 0x86
    CMD_SET_KEY = 0x87
    CMD_GET_STATUS = 0x88
    CMD_SET_RADIO_ID = 0x89


class Mode(IntEnum):
    IDLE = 0
    CAPTURE = 1
    ED_SWEEP = 2
    CAPTURE_PLUS_ED = 3


FLAG_CRC_OK = 0x01
FLAG_PROMISCUOUS = 0x02
FLAG_FRAME_PENDING = 0x04
FLAG_WAS_ACKED = 0x08


def crc16(data: bytes) -> int:
    """CRC-16/CCITT-FALSE: poly 0x1021, init 0xFFFF, no reflection."""
    crc = 0xFFFF
    for b in data:
        crc ^= b << 8
        for _ in range(8):
            crc = ((crc << 1) ^ 0x1021) & 0xFFFF if (crc & 0x8000) else (crc << 1) & 0xFFFF
    return crc


def encode(msg_type: int, payload: bytes = b"") -> bytes:
    """Build a complete framed message."""
    if len(payload) > MAX_PAYLOAD:
        raise ValueError("payload too large")
    header_body = struct.pack("<BBH", PROTO_VER, msg_type, len(payload)) + payload
    crc = crc16(header_body)
    return bytes([MAGIC0, MAGIC1]) + header_body + struct.pack("<H", crc)


# --- command builders ------------------------------------------------------

def cmd_set_channel(ch: int) -> bytes:
    return encode(Msg.CMD_SET_CHANNEL, bytes([ch]))


def cmd_set_mode(mode: int) -> bytes:
    return encode(Msg.CMD_SET_MODE, bytes([mode]))


def cmd_start() -> bytes:
    return encode(Msg.CMD_START)


def cmd_stop() -> bytes:
    return encode(Msg.CMD_STOP)


def cmd_set_hop(hop_mask: int, dwell_ms: int) -> bytes:
    return encode(Msg.CMD_SET_HOP, struct.pack("<IH", hop_mask, dwell_ms))


def cmd_ed_scan(hop_mask: int, dwell_ms: int) -> bytes:
    return encode(Msg.CMD_ED_SCAN, struct.pack("<IH", hop_mask, dwell_ms))


def cmd_set_key(key: bytes | None) -> bytes:
    if key is None:
        return encode(Msg.CMD_SET_KEY, bytes([0]))
    if len(key) != 16:
        raise ValueError("network key must be 16 bytes")
    return encode(Msg.CMD_SET_KEY, bytes([1]) + key)


def cmd_get_status() -> bytes:
    return encode(Msg.CMD_GET_STATUS)


def all_channels_mask() -> int:
    return sum(1 << (ch - CHANNEL_MIN) for ch in range(CHANNEL_MIN, CHANNEL_MAX + 1))


# --- parsed message types --------------------------------------------------

@dataclass
class CapturedFrame:
    radio_id: int
    channel: int
    rssi: int
    lqi: int
    flags: int
    timestamp: int  # microseconds
    mpdu: bytes     # raw 802.15.4 MPDU, includes FCS

    @property
    def crc_ok(self) -> bool:
        return bool(self.flags & FLAG_CRC_OK)


@dataclass
class EdResult:
    radio_id: int
    channel: int
    timestamp: int
    ed_dbm: int
    sweep_id: int


@dataclass
class Status:
    radio_id: int
    mode: int
    channel: int
    hop_mask: int
    hop_dwell_ms: int
    uptime_s: int
    captured: int
    dropped_crc: int
    dropped_buf: int
    fw_major: int
    fw_minor: int


@dataclass
class LogLine:
    text: str


@dataclass
class Incident:
    raw: str          # UTF-8 JSON as emitted by the device
    data: dict        # parsed (empty dict if it failed to parse)


@dataclass
class Ack:
    cmd_type: int
    result: int


def parse_payload(msg_type: int, p: bytes):
    """Decode a payload into one of the dataclasses above (or None if unknown)."""
    if msg_type == Msg.CAPTURED_FRAME:
        radio_id, channel, rssi, lqi, flags = struct.unpack_from("<BBbBB", p, 0)
        (timestamp,) = struct.unpack_from("<Q", p, 5)
        mpdu_len = p[13]
        mpdu = p[14:14 + mpdu_len]
        return CapturedFrame(radio_id, channel, rssi, lqi, flags, timestamp, mpdu)
    if msg_type == Msg.ED_RESULT:
        radio_id, channel = p[0], p[1]
        (timestamp,) = struct.unpack_from("<Q", p, 2)
        ed_dbm = struct.unpack_from("<b", p, 10)[0]
        (sweep_id,) = struct.unpack_from("<H", p, 11)
        return EdResult(radio_id, channel, timestamp, ed_dbm, sweep_id)
    if msg_type == Msg.STATUS:
        (radio_id, mode, channel, hop_mask, hop_dwell_ms, uptime_s,
         captured, dropped_crc, dropped_buf, fw_major, fw_minor) = struct.unpack_from(
            "<BBBIHIIIIBB", p, 0)
        return Status(radio_id, mode, channel, hop_mask, hop_dwell_ms, uptime_s,
                      captured, dropped_crc, dropped_buf, fw_major, fw_minor)
    if msg_type == Msg.LOG:
        return LogLine(p.decode("utf-8", "replace"))
    if msg_type == Msg.INCIDENT:
        raw = p.decode("utf-8", "replace")
        try:
            return Incident(raw, json.loads(raw))
        except ValueError:
            return Incident(raw, {})
    if msg_type == Msg.ACK:
        return Ack(p[0], p[1])
    return None


class StreamDecoder:
    """Incremental frame decoder mirroring the C zb_decoder_t state machine."""

    def __init__(self) -> None:
        self._buf = bytearray()

    def feed(self, data: bytes):
        """Feed raw bytes; yield (msg_type, payload_bytes) for each valid frame."""
        self._buf.extend(data)
        out = []
        while True:
            frame = self._try_extract()
            if frame is None:
                break
            out.append(frame)
        return out

    def _try_extract(self):
        b = self._buf
        # find magic
        i = b.find(bytes([MAGIC0, MAGIC1]))
        if i < 0:
            # keep at most one trailing byte (could be a partial magic)
            if len(b) > 1:
                del b[:-1]
            return None
        if i > 0:
            del b[:i]
        if len(b) < 6:
            return None
        ver, mtype, plen = struct.unpack_from("<BBH", b, 2)
        if ver != PROTO_VER or plen > MAX_PAYLOAD:
            del b[:2]  # bad header, skip past this magic
            return None
        total = 6 + plen + 2
        if len(b) < total:
            return None
        body = bytes(b[2:6 + plen])
        (crc_rx,) = struct.unpack_from("<H", b, 6 + plen)
        del b[:total]
        if crc16(body) != crc_rx:
            return None
        return (mtype, body[4:])

File: tether/internal/proto/proto.go
// Package proto implements the device<->host wire framing (v1).
// Port of host/zbsniff/proto.py and firmware/src/core/codec.c — keep in sync
// with /protocol/framing.md.
package proto

import (
	"encoding/binary"
	"encoding/json"
)

const (
	Magic0 = 0x5A
	Magic1 = 0xBE
	Ver    = 1

	MaxPayload = 2048
	ChannelMin = 11
	ChannelMax = 26
)

// Message types
const (
	MsgCapturedFrame = 0x01
	MsgEdResult      = 0x02
	MsgStatus        = 0x03
	MsgLog           = 0x04
	MsgAck           = 0x05
	MsgIncident      = 0x06
	MsgProbeResult   = 0x07
	MsgOtaStatus     = 0x08

	CmdSetChannel = 0x81
	CmdSetMode    = 0x82
	CmdStart      = 0x83
	CmdStop       = 0x84
	CmdSetHop     = 0x85
	CmdEdScan     = 0x86
	CmdSetKey     = 0x87
	CmdGetStatus  = 0x88
	CmdSetRadioID = 0x89
	CmdProbe      = 0x8A
	CmdOtaBegin   = 0x8B
	CmdOtaData    = 0x8C
	CmdOtaEnd     = 0x8D
	CmdOtaAbort   = 0x8E
	CmdBeaconReq  = 0x8F
	CmdTxRaw      = 0x90

	// Relayed satellite control (target(1) + inner args, stripped and forwarded
	// over SPI by the primary) — see docs on CMD_SET_CHANNEL/CMD_START/CMD_STOP.
	CmdSatSetChannel = 0x91
	CmdSatStart      = 0x92
	CmdSatStop       = 0x93
	CmdSatRelay      = 0x94
)

// OTA states (MsgOtaStatus.State).
const (
	OtaIdle = iota
	OtaReceiving
	OtaWriting
	OtaVerifying
	OtaOK
	OtaError
)

// Capture modes
const (
	ModeIdle          = 0
	ModeCapture       = 1
	ModeEdSweep       = 2
	ModeCapturePlusEd = 3
)

// CAPTURED_FRAME flags
const (
	FlagCRCOK       = 0x01
	FlagPromiscuous = 0x02
	FlagFramePend   = 0x04
	FlagWasAcked    = 0x08
)

// CRC16 computes CRC-16/CCITT-FALSE (poly 0x1021, init 0xFFFF).
func CRC16(data []byte) uint16 {
	crc := uint16(0xFFFF)
	for _, b := range data {
		crc ^= uint16(b) << 8
		for i := 0; i < 8; i++ {
			if crc&0x8000 != 0 {
				crc = (crc << 1) ^ 0x1021
			} else {
				crc <<= 1
			}
		}
	}
	return crc
}

// Encode builds a complete framed message.
func Encode(msgType byte, payload []byte) []byte {
	body := make([]byte, 4+len(payload))
	body[0] = Ver
	body[1] = msgType
	binary.LittleEndian.PutUint16(body[2:], uint16(len(payload)))
	copy(body[4:], payload)
	crc := CRC16(body)
	out := make([]byte, 2+len(body)+2)
	out[0] = Magic0
	out[1] = Magic1
	copy(out[2:], body)
	binary.LittleEndian.PutUint16(out[2+len(body):], crc)
	return out
}

// Command builders
func CmdSetChannelMsg(ch byte) []byte { return Encode(CmdSetChannel, []byte{ch}) }
func CmdSetModeMsg(m byte) []byte     { return Encode(CmdSetMode, []byte{m}) }
func CmdStartMsg() []byte             { return Encode(CmdStart, nil) }
func CmdStopMsg() []byte              { return Encode(CmdStop, nil) }
func CmdGetStatusMsg() []byte         { return Encode(CmdGetStatus, nil) }

// AllChannelsMask returns the bitmask of channels 11..26 (bit i = channel 11+i).
func AllChannelsMask() uint32 {
	var m uint32
	for c := ChannelMin; c <= ChannelMax; c++ {
		m |= 1 << uint(c-ChannelMin)
	}
	return m
}

// CmdEdScanMsg requests a one-shot energy-detect sweep of hopMask (dwell ms/channel).
func CmdEdScanMsg(hopMask uint32, dwellMs uint16) []byte {
	p := make([]byte, 6)
	binary.LittleEndian.PutUint32(p[0:], hopMask)
	binary.LittleEndian.PutUint16(p[4:], dwellMs)
	return Encode(CmdEdScan, p)
}

// CmdSetHopMsg configures channel hopping (dwellMs=0 disables / pins the channel).
func CmdSetHopMsg(hopMask uint32, dwellMs uint16) []byte {
	p := make([]byte, 6)
	binary.LittleEndian.PutUint32(p[0:], hopMask)
	binary.LittleEndian.PutUint16(p[4:], dwellMs)
	return Encode(CmdSetHop, p)
}

// CmdSetRadioIDMsg provisions a dongle's radio-id (persisted to its NVS).
func CmdSetRadioIDMsg(id byte) []byte { return Encode(CmdSetRadioID, []byte{id}) }

// OTA command builders. target 0 = the tethered C6, 1..3 = a satellite.
func CmdOtaBeginMsg(target byte, total, crc32 uint32) []byte {
	p := make([]byte, 9)
	p[0] = target
	binary.LittleEndian.PutUint32(p[1:], total)
	binary.LittleEndian.PutUint32(p[5:], crc32)
	return Encode(CmdOtaBegin, p)
}
// CmdOtaDataMsg's wire format is deliberately NOT versioned/extended lightly:
// this message is parsed by whatever firmware is CURRENTLY RUNNING on the
// target, which is exactly the firmware an OTA update is trying to replace.
// A field added here only helps once new firmware understanding it is
// already installed — but the only way to install it is a transfer using
// THIS message, parsed by the OLD firmware. Changing the format silently
// breaks that transfer for every device still running anything older,
// with no way to recover except a physical reflash. (Learned the hard way:
// an added content-checksum field shifted every chunk's data by 4 bytes as
// seen by old firmware, corrupting the image from byte zero — every
// subsequent failure looked like a deeper bug until this was traced back.)
func CmdOtaDataMsg(target byte, offset uint32, chunk []byte) []byte {
	p := make([]byte, 5+len(chunk))
	p[0] = target
	binary.LittleEndian.PutUint32(p[1:], offset)
	copy(p[5:], chunk)
	return Encode(CmdOtaData, p)
}
func CmdOtaEndMsg(target byte) []byte   { return Encode(CmdOtaEnd, []byte{target}) }
func CmdOtaAbortMsg(target byte) []byte { return Encode(CmdOtaAbort, []byte{target}) }

// Relayed satellite control builders. target is the satellite's radio-id (1..3).
func CmdSatSetChannelMsg(target, ch byte) []byte { return Encode(CmdSatSetChannel, []byte{target, ch}) }
func CmdSatStartMsg(target byte) []byte          { return Encode(CmdSatStart, []byte{target}) }
func CmdSatStopMsg(target byte) []byte           { return Encode(CmdSatStop, []byte{target}) }

// CmdSatRelayMsg wraps a fully-encoded inner command frame so the primary
// forwards it verbatim over SPI to satellite `target` (1..3). This is how any
// host->device command (mode, ED, hop, probe, beacon, raw TX) reaches a
// satellite — see CMD_SAT_RELAY in proto.h.
func CmdSatRelayMsg(target byte, inner []byte) []byte {
	return Encode(CmdSatRelay, append([]byte{target}, inner...))
}

// CmdBeaconReqMsg asks the radio to transmit an 802.15.4 beacon request on the
// current channel; coordinators/routers reply with a beacon revealing their PAN.
func CmdBeaconReqMsg() []byte { return Encode(CmdBeaconReq, nil) }

// CmdTxRawMsg transmits a host-built raw MPDU (the radio appends the FCS).
func CmdTxRawMsg(mpdu []byte) []byte { return Encode(CmdTxRaw, mpdu) }

// CmdProbeMsg requests an active MAC probe of target (short addr) on pan.
func CmdProbeMsg(target, pan uint16) []byte {
	p := make([]byte, 4)
	binary.LittleEndian.PutUint16(p[0:], target)
	binary.LittleEndian.PutUint16(p[2:], pan)
	return Encode(CmdProbe, p)
}

func CmdSetKeyMsg(key []byte) []byte {
	if len(key) != 16 {
		return Encode(CmdSetKey, []byte{0})
	}
	return Encode(CmdSetKey, append([]byte{1}, key...))
}

// Parsed message payloads
type CapturedFrame struct {
	RadioID   byte
	Channel   byte
	RSSI      int8
	LQI       byte
	Flags     byte
	Timestamp uint64
	MPDU      []byte
}

type EdResult struct {
	RadioID byte
	Channel byte
	EdDBm   int8
	SweepID uint16
}

type Status struct {
	RadioID    byte
	Mode       byte
	Channel    byte
	HopMask    uint32
	HopDwellMs uint16
	Captured   uint32
	DroppedBuf uint32
	UptimeS    uint32
	FWMajor    byte
	FWMinor    byte
}

type Incident struct {
	Raw  string
	Data map[string]any
}

// ProbeResult is the outcome of an active MAC probe.
type ProbeResult struct {
	RadioID byte
	Target  uint16
	Acked   bool
	RSSI    int8
	LQI     byte
}

// OtaStatus is an OTA progress report.
type OtaStatus struct {
	Target   byte
	State    byte
	Received uint32
	Total    uint32
	Err      byte
}

// ParsePayload decodes a payload by message type. Returns nil for unknown types.
func ParsePayload(msgType byte, p []byte) any {
	switch msgType {
	case MsgCapturedFrame:
		if len(p) < 14 {
			return nil
		}
		mlen := int(p[13])
		if 14+mlen > len(p) {
			return nil
		}
		return &CapturedFrame{
			RadioID:   p[0],
			Channel:   p[1],
			RSSI:      int8(p[2]),
			LQI:       p[3],
			Flags:     p[4],
			Timestamp: binary.LittleEndian.Uint64(p[5:13]),
			MPDU:      append([]byte(nil), p[14:14+mlen]...),
		}
	case MsgEdResult:
		if len(p) < 13 {
			return nil
		}
		return &EdResult{
			RadioID: p[0], Channel: p[1],
			EdDBm:   int8(p[10]),
			SweepID: binary.LittleEndian.Uint16(p[11:13]),
		}
	case MsgStatus:
		if len(p) < 27 {
			return nil
		}
		return &Status{
			RadioID:    p[0],
			Mode:       p[1],
			Channel:    p[2],
			HopMask:    binary.LittleEndian.Uint32(p[3:7]),
			HopDwellMs: binary.LittleEndian.Uint16(p[7:9]),
			UptimeS:    binary.LittleEndian.Uint32(p[9:13]),
			Captured:   binary.LittleEndian.Uint32(p[13:17]),
			DroppedBuf: binary.LittleEndian.Uint32(p[21:25]),
			FWMajor:    p[25],
			FWMinor:    p[26],
		}
	case MsgIncident:
		var m map[string]any
		_ = json.Unmarshal(p, &m)
		return &Incident{Raw: string(p), Data: m}
	case MsgProbeResult:
		if len(p) < 6 {
			return nil
		}
		return &ProbeResult{
			RadioID: p[0],
			Target:  binary.LittleEndian.Uint16(p[1:3]),
			Acked:   p[3] != 0,
			RSSI:    int8(p[4]),
			LQI:     p[5],
		}
	case MsgOtaStatus:
		if len(p) < 11 {
			return nil
		}
		return &OtaStatus{
			Target:   p[0],
			State:    p[1],
			Received: binary.LittleEndian.Uint32(p[2:6]),
			Total:    binary.LittleEndian.Uint32(p[6:10]),
			Err:      p[10],
		}
	}
	return nil
}

// StreamDecoder reassembles framed messages from a byte stream.
type StreamDecoder struct {
	buf []byte
}

// Frame is a decoded (type, payload) pair.
type Frame struct {
	Type    byte
	Payload []byte
}

// Feed appends data and returns any complete, CRC-valid frames.
func (d *StreamDecoder) Feed(data []byte) []Frame {
	d.buf = append(d.buf, data...)
	var out []Frame
	i := 0
	for {
		// find magic
		for i+1 < len(d.buf) && !(d.buf[i] == Magic0 && d.buf[i+1] == Magic1) {
			i++
		}
		if i+6 > len(d.buf) {
			break
		}
		if d.buf[i+2] != Ver {
			i++
			continue
		}
		plen := int(binary.LittleEndian.Uint16(d.buf[i+4:]))
		if plen > MaxPayload {
			i++
			continue
		}
		total := 6 + plen + 2
		if i+total > len(d.buf) {
			break
		}
		crcRx := binary.LittleEndian.Uint16(d.buf[i+6+plen:])
		if CRC16(d.buf[i+2:i+6+plen]) == crcRx {
			payload := append([]byte(nil), d.buf[i+6:i+6+plen]...)
			out = append(out, Frame{Type: d.buf[i+3], Payload: payload})
			i += total
		} else {
			i++
		}
	}
	d.buf = append([]byte(nil), d.buf[i:]...)
	return out
}
